BuildScript/BuildTasks/TestConfigurationsFactory.cs

```diff
@@ -0,0 +1,220 @@
+      }
+    }
+
+    private TestConfiguration CreateTestConfiguration (
+        string testAssemblyFullPath,
+        IReadOnlyCollection<string> configurationItems,
+        string testingSetupBuildFile)
+    {
+      var browser = GetBrowser (configurationItems);
+      var databaseSystem = GetDatabaseSystem (configurationItems);
+      var platform = GetPlatform (configurationItems);
+      var targetRuntime = GetTargetRuntime (configurationItems);
+      var executionRuntime = GetExecutionRuntime (configurationItems);
+      var configurationID = GetConfigurationID (configurationItems);
+      var testAssemblyFileName = Path.GetFileName (testAssemblyFullPath);
+      var targetRuntimeInitialFormat = TargetRuntimeConverter.ToInitialFormat (targetRuntime);
+      var executionRuntimeKey = _supportedExecutionRuntimes.GetKeysByValue (executionRuntime).First();
+
+      return new TestConfiguration ( //todo exclude and include categories need to be here too
+          id: $"{testAssemblyFileName}_{browser}+{databaseSystem}+{platform}+{executionRuntimeKey}+{targetRuntimeInitialFormat}+{configurationID}",
+          browser: browser,
+          isWebTest: !string.Equals (browser, MetadataValueConstants.NoBrowser, StringComparison.OrdinalIgnoreCase),
+          databaseSystem: databaseSystem,
+          isDatabaseTest: !string.Equals (databaseSystem, MetadataValueConstants.NoDB, StringComparison.OrdinalIgnoreCase),
+          configurationID: configurationID,
+          use32Bit: string.Equals (platform, "x86", StringComparison.OrdinalIgnoreCase),
+          platform: platform,
+          executionRuntime: executionRuntime,
+          useDocker: ShouldUseDocker (executionRuntime),
+          testAssemblyFileName: testAssemblyFileName,
+          testAssemblyFullPath: testAssemblyFullPath,
+          testAssemblyDirectoryName: Path.GetDirectoryName (testAssemblyFullPath),
+          testSetupBuildFile: testingSetupBuildFile,
+          targetRuntime: targetRuntime,
+          excludeCategories: GetExcludeCategories (configurationItems),
+          includeCategories: GetIncludeCategories (configurationItems));
+    }
+
+    private bool ShouldUseDocker (string executionRuntime)
+    {
+      return !string.Equals (executionRuntime, MetadataValueConstants.LocalMachine, StringComparison.OrdinalIgnoreCase)
+             && !string.Equals (executionRuntime, MetadataValueConstants.EnforcedLocalMachine, StringComparison.OrdinalIgnoreCase);
+    }
+
+    private IReadOnlyCollection<string> GetDuplicateItems (IReadOnlyCollection<string> configurationItems)
+    {
+      return configurationItems
+          .GroupBy (x => x)
+          .Where (group => group.Count() > 1)
+          .Select (group => group.Key)
+          .ToArray();
+    }
+
+    private string GetConfigurationID (IReadOnlyCollection<string> configurationItems)
+    {
+      var configurationID = _supportedConfigurationIDs.SingleOrDefault (configurationItems);
```

Comment:
What if we return an IEnumerable and add the SingleOrDefualt as an extension method? that way, we can also do the null-check as an extension method like we do in EnumerableUtility

Reply:
I believe that would just be additional overhead without making the code more readable/maintainable, especially considering it's already pretty short and readable at this point

Reply:
Am not convinced that's true. To me, it just feels wrong to add SingeOrDefault (including overloads for the exception) to the actual type instead of as an extension method. Especially, if that's the only API on the collection type.

Reply:
No, a little misunderstanding. I meant encapsulating the exception throw at all wouldn't be too helpful, but I did it anyway, tell me if it's what you imagined (after re-request)

Reply:
Yes, but only for "Single" or "First", not "GetExecutionRuntime "